src/Processors/Transforms/Streaming/JoinTransform.cpp

```diff
@@ -278,21 +269,18 @@ inline void JoinTransform::joinBidirectionally(Chunks chunks)
         auto block = input_ports_with_data[i].input_port->getHeader().cloneWithColumns(chunks[i].detachColumns());
         auto retracted_block = std::invoke(join_funcs[i], join.get(), block);
 
+        /// First emit retracted block
+        auto retracted_block_rows = retracted_block.rows();
+        if (retracted_block_rows)
         {
-            std::scoped_lock lock(mutex);
-            /// First emit retracted block
-            auto retracted_block_rows = retracted_block.rows();
-            if (retracted_block_rows)
-            {
-                /// Don't watermark this block. We can concat retracted / result blocks or use avoid watermarking
-                auto chunk_ctx = ChunkContext::create();
-                chunk_ctx->setConsecutiveDataFlag();
-                output_chunks.emplace_back(retracted_block.getColumns(), retracted_block_rows, nullptr, std::move(chunk_ctx));
-            }
-
-            if (block.rows())
-                output_chunks.emplace_back(block.getColumns(), block.rows());
+            /// Don't watermark this block. We can concat retracted / result blocks or use avoid watermarking
```

Comment:
```
/// Don't watermark retracted chunk since we like the retracted chunk and the following result chunk
/// to process in a consecutive way (for example, avoid emitting result right after processing retracted chunk
/// but without processing the following result chunk. This will prevent transitive emit result we don't like 
/// to have usually. 
/// To have retracted chunk / result chunk processed consecutively, we can either concat them into one bigger
/// chunk or use `consecutive` flag which we use here.
```